Libraries/LibPCIDB/Database.cpp

```diff
@@ -0,0 +1,205 @@
+    if (!parent)
+        return {};
+
+    auto res = parent->subclasses.get(subclass_id);
+    if (res.has_value())
+        return res.value();
+    return {};
+}
+
+RefPtr<ProgrammingInterface> Database::get_programming_interface(u8 class_id, u8 subclass_id, u8 programming_interface_id) const
+{
+    auto parent = get_subclass(class_id, subclass_id);
+    if (!parent)
+        return {};
+
+    auto res = parent->programming_interfaces.get(programming_interface_id);
+    if (res.has_value())
+        return res.value();
+    return {};
+}
+
+u8 parse_hex_digit(char digit)
+{
+    if (digit >= '0' && digit <= '9')
+        return digit - '0';
+    ASSERT(digit >= 'a' && digit <= 'f');
+    return 10 + (digit - 'a');
+}
+
+template<typename T>
+T parse_hex(StringView str, int count)
+{
+    ASSERT(str.length() >= count);
+
+    T res = 0;
+    for (int i = 0; i < count; i++)
+        res = (res << 4) + parse_hex_digit(str[i]);
+
+    return res;
+}
+
+int Database::init()
+{
+    if (m_ready)
+        return 0;
+
+    if (!m_file.is_valid())
+        return -1;
+
+    m_view = StringView((char*)m_file.pointer(), m_file.size());
+
+    ParseMode mode = ParseMode::UnknownMode;
+
+    Vendor* current_vendor = nullptr;
+    Device* current_device = nullptr;
+    Class* current_class = nullptr;
+    Subclass* current_subclass = nullptr;
+
+    auto commit_device = [&]() {
+        if (current_device && current_vendor)
+            current_vendor->devices.set(current_device->id, *current_device);
+        current_device = nullptr;
+    };
+
+    auto commit_vendor = [&]() {
+        commit_device();
+        if (current_vendor)
+            m_vendors.set(current_vendor->id, *current_vendor);
+        current_vendor = nullptr;
+    };
+
+    auto commit_subclass = [&]() {
+        if (current_subclass && current_class)
+            current_class->subclasses.set(current_subclass->id, *current_subclass);
+        current_subclass = nullptr;
+    };
+
+    auto commit_class = [&]() {
+        commit_subclass();
+        if (current_class)
+            m_classes.set(current_class->id, *current_class);
+        current_class = nullptr;
+    };
+
+    auto commit_all = [&]() {
+        commit_vendor();
+        commit_class();
+    };
+
+    auto lines = m_view.split_view('\n');
+
+    for (auto& line : lines) {
+        if (line.length() < 2 || line[0] == '#')
+            continue;
+
+        if (line[0] == 'C') {
+            mode = ParseMode::ClassMode;
+            commit_all();
+        } else if ((line[0] >= '0' && line[0] <= '9') || (line[0] >= 'a' && line[0] <= 'f')) {
+            mode = ParseMode::VendorMode;
+            commit_all();
+        } else if (line[0] != '\t') {
+            mode = ParseMode::UnknownMode;
+            continue;
+        }
+
+        switch (mode) {
+        case ParseMode::VendorMode:
+            if (line[0] != '\t') {
+                commit_vendor();
+                current_vendor = new Vendor;
```

Comment:
Assuming we've made these into `RefPtr`s:
```suggestion
                current_vendor = adopt(*new Vendor);
```